### PR TITLE
Add magazines to many guns [RDY]

### PR DESCRIPTION
**Pull request by mugling**
@Rivet-the-Zombie, as discussed in #15007 here is a PR combining #14995, #15002, #15003, #15009 with all of the merge conflicts resolved. To keep the size down I've left #15007 out for now as the `8x40` ammo type isn't used in bore conversions so no need to specify `alternatives`.

No new content is added but some item id's are changed to be consistent and accordingly there is a small patch (3a1ec1d) for the mshock tileset at the end.

@chaosvolt a lot of the specified `alternatives` in #15002 are omitted because they will never be used due to a lack of bore conversions. Eg. there is no bore conversion changing `9mm` in to `9x18`. Such a gunmod should probably exist to offset the intentional rarity of magazines for such weapons. We probably need to increase the penalties for bore mods also as they are now **much** more useful for situations where you haven't yet collected enough magazines of a type.

### CODE REVIEW

File: data/json/items/magazines.json [hunk 1]
```diff
@@ -65,14 +352,314 @@
...
+        "color" : "ltgray",
+        "description" : "A 50-round drum magazine for the Thompson submachine gun.  Made famous by gangsters, but the military preferred standard 20 and 30-round magazines due to greater reliablity.",
+        "material" : "steel",
+        "volume" : 4,
+        "weight" : 500,
+        "ammo_type" : "45",
+        "capacity" : 50,
+        "reliability" : 7,
+        "reload_time" : 150,
+        "flags": [ "BULKY_MAG" ]
```
Comment (chaosvolt) [inline data/json/items/magazines.json]: No conversion given for 9mm?

**Comment by chaosvolt** [comment]
Ah, thank you for pointing that out. I had assumed that conversions for it could be added later on, after establishing the initial foundation for its compatibility.

And I'm guessing this includes my additions as well, so I can close my PR for it.

File: data/json/items/magazines.json [hunk 2]
```diff
@@ -65,14 +352,314 @@
...
+        "reliability" : 7,
+        "reload_time" : 120,
+        "flags": [ "COMPACT_MAG" ],
+        "alternatives" : [
+            [ "223", [ "stanag30", "stanag10", "survivor223mag" ] ]
+        ]
+    },
+    {
+        "type" : "MAGAZINE",
+        "id" : "belt308",
```
Comment (chaosvolt) [inline data/json/items/magazines.json]: Didn't also include the DM1 reusable belt I'd added?

Comment (mugling) [inline data/json/items/magazines.json]: Yes, but I'm still reworking ammo belts so I've left this out for now

Comment (chaosvolt) [inline data/json/items/magazines.json]: Technically a reusable belt would be effectively just another magazine though, though potentially one that's a pain in the ass to reload. :V

**Comment by mugling** [comment]
Although 5228eca includes disassembly recipes this PR will **not ** include crafting recipes for ammo belts

**Comment by chaosvolt** [comment]
Ah, interesting. Will be useful for unloading the belts.

Though the eventually scarcity of ammo belts will be a reason for non-disintegrating belts to exist. XP

**Comment by mugling** [comment]
As of d099de3 `COMPACT_MAG` has become `MAG_COMPACT` and `BULKY_MAG` has become `MAG_BULKY`. Documentation fixes will be included as part of #15099.

This change is for consistency when implementing `MAG_DESTROY` and `MAG_EJECT`

**Comment by kevingranade** [comment]
Thomson hasn't been adjusted, so it won't accept magazines.  That was literally the first thing I checked, so it's likely there are other mismatches.
Now that I've had a chance to see it in action, I'm unclear on why the alternate magazine list is applied to the magazine instead of to the gun.  It's the gun that gets converted, so it's the object that should "know" what the alternatives are.  If you're trying to minimize repetition, it looks like that's going to be a wash as well, it looks like there are going to be similar numbers of magazines as guns that accept magazines.

**Comment by chaosvolt** [comment]
> Thomson hasn't been adjusted, so it won't accept magazines. That was literally the first thing I checked, so it's likely there are other mismatches.

I added magazines to the Thompson in https://github.com/CleverRaven/Cataclysm-DDA/pull/15002 at least, so it wasn't my mistake for once. ;w;

**Comment by kevingranade** [comment]
Looks like it's just the Thompson, there's no gun that uses belt308, but it's used as an alternate, should that ever actually happen?  I added a consistency check to insure all magazines that are used are defined and vice versa, but it only checks magazines vs guns, not alternates.  The problem with including alternates is it can introduce loops, where there are 2 magazine types that list each other as alternates, but no guns use them.

**Comment by kevingranade** [comment]
@mugling I PRed the consistency check to you.

**Comment by chaosvolt** [comment]
> there's no gun that uses belt308, but it's used as an alternate, should that ever actually happen?

Again, I added the belt308 to the M60 and M240. @mugling done derped. o3o

**Comment by mugling** [comment]
@kevingranade nice, catches the unused magazines

**Comment by mugling** [comment]
> there's no gun that uses belt308, but it's used as an alternate, should that ever actually happen?

No it's an omission on my part

**Comment by mugling** [comment]
> I'm unclear on why the alternate magazine list is applied to the magazine instead of to the gun. It's the gun that gets converted, so it's the object that should "know" what the alternatives are.

I agree it's semantically better the other way around but I'm wary about changing it as part of this PR especially as this and #15081 are blocking further PRs relating to magazines.

Would it acceptable if we fix the remaining obstacles (in particular #15081) and I revisit the `alternatives` issue later as it will require changes to `item::magazine_compatible` and probably further. 

Any JSON changes (whether now or later) would themselves would be trivial.

**Comment by chaosvolt** [comment]
Right. Now I'm wondering whether it was that bright an idea to compile all the magazine additions into a single hefty PR.

**Comment by mugling** [comment]
@chaosvolt the problem is inter-dependencies and also that I'm still working on the implementation details. For example d099de3 renamed some flags and a5e7c81 added others. Once the implementation is concrete things should be easier. Feel free to submit a PR against this branch for the M60 and M240 though as they're otherwise good to go.

**Comment by chaosvolt** [comment]
Hmm, true. Not sure how best to handle it then. Already being a bit hectic with the number of PRs out now. :V

**Comment by kevingranade** [comment]
Swapping how it's defined later is fine, just wanted to raise the issue.

**Comment by mugling** [comment]
Ok so I think that's all the outstanding problems

**Comment by Malkeus** [comment]
Glurk!?
https://i.imgur.com/WchvES0.png

**Comment by chaosvolt** [comment]
Ohgodwhat. ;w;

**Comment by Malkeus** [comment]
Am I the only one getting a wall of red when entering a game ? The save is
only a month old.

**Comment by mugling** [comment]
Can't reproduce. Does it affect new games or just old saves?

**Comment by chaosvolt** [comment]
Yeah, start up a new game to test whether it isn't just a save compatibility issue, which I assumed could happen.

**Comment by mugling** [comment]
Although I did test for that and I presume so did others? Can you try building master and checking against that?

**Comment by Rivet-the-Zombie** [comment]
I can't reproduce this one. Maybe it's mod related?

**Comment by chaosvolt** [comment]
Hmm. I'll grab the latest build in a bit and transfer Snips over to it. I'd already previously had fun with it because I'd used the original build it was from to test my initial magazine PR, which led to a brief error after I later transferred to a fresh build that no longer had Sten magazines. XP

**Comment by Malkeus** [comment]
You're right @Rivet-the-Zombie , it appears to be a conflict with the Guns/Equipment/Weapons Expansion Mode. That and artyom's need to be updated to the new magazine system anyway, I'll remove them for now. I thought item-groups worked in an additive fashion though, ie. anytime a mod had a matching item_group id the data would be appended to the core game version. Is that wrong?

**Comment by chaosvolt** [comment]
Ah, odd. And itemgroups should be additive, yeah. If it was override-based like monstergroups are, a LOT of mods would be royally fucked. :V

**Comment by Malkeus** [comment]
Maybe it's the modified guns that are doing it. Either way, they are gone til they get straightened out. I hope my laser sniper rifle isn't from one of those...

Starting a new unmodded world gives no errors, so everything is cool on your end @mugling.

**Comment by Malkeus** [comment]
Oi, so many debug messages about missing items :(

**Comment by chaosvolt** [comment]
It must've been the mods, as transferred a save over without issue.

**Comment by Malkeus** [comment]
Agreed. I haven't gone digging to confirm, but I believe one or both of
those mods alters core game guns to give them magazines. This PR alters the
same files for the same reason, then the mod overrides it, leading to the
error. They will need to be updated. I did find a nifty bug with gunmods
requiring an unloaded weapon and ups powered weapons that can't be unloaded
after removing the mods from my save.

**Comment by chaosvolt** [comment]
Hmm. Definitely peculiar. o.o

**Comment by kevingranade** [comment]
I suspect the mid is re-instituting the old version of the gun, which means
it doesn't reference the magazines.

**Comment by chaosvolt** [comment]
I think that was what Malkeus already suspected as well, that must be it.

**Comment by mugling** [comment]
I don't have a forum account but it might be worth a post to the **Latest experimental features** thread stating that support for magazines is in mainline and which mods are currently incompatible.

**Comment by chaosvolt** [comment]
Doh. In addition likely mention it in the threads for those mods, though by now others might have.

**Comment by Malkeus** [comment]
I already dropped a post into both about this bug, I don't think I
mentioned the new magazine code specifically though. When they investigate
the bug it should become apparent...

**Comment by chaosvolt** [comment]
Ah, hopefully.

**Pull request merged**